### ISSUE TITLE
PTX alignment mismatch for complex outputs of function b/w NVCC and Clang

### ISSUE THREAD

**Issue by rashedmyt** [llvm/llvm-project#52840]
https://cuda.godbolt.org/z/aaeshann8

Notice the alignment mismatch for the output in the function declaration of `add`.

**Comment by Artem-B** [comment]
This is an interesting one, even if it appears to be cosmetic.

The concept of alignment applies to loads/stores/pointers and not to the types themselves. LLVM has no idea that clang has `__align__(16)` applied to a type.

On the other hand, PTX pretends that return values are passed via loads/stores to/from a fictional `.param` memory.
At the time we generate those loads/stores we do not have the alignment info and therefore can only rely on the alignment specified by the ABI.

This, theoretically, results in a less optimial PTX as we would not generate vectorized loads and stores.
On the other hand, `.param` memory is not real. AFAICT, the value is actually returned via registers, so load/store vectorization and the notional alignment of the result does not matter in practice. 
https://godbolt.org/z/hdWssbK74 -- call of `func` only uses registers, if you look at the SASS disassembly.

Do you have an actual example where this mismatch in declared alignment of the return value is a problem correctness- or performance-wise? As far as I can tell this is a purely cosmetic issue.

**Comment by rashedmyt** [comment]
Though I cannot provide any code, Below is a brief overview of the use case.

The software I am modifying is a JIT compiler which creates LLVM IR from the source language. The compiler then uses LLVM NVPTX backend for PTX generation which is then compiled to SASS by CUDA driver API (_cuLink*_, _cuModule*_). We also have separate CUDA device code compiled by NVCC to a FATBINARY which is linked with the PTX generated using the driver API. The CUDA device code uses certain vector types mentioned in [CUDA programming guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#vector-types). To generate corresponding function declarations we use LLVM vector types to match the CUDA vector types.

Using NVCC has its pro's and con's and we decided on compiling the CUDA device code to LLVM bitcode and use it directly in the PTX generation phase giving chance for many optimizations (inlining etc). However, when generating SASS from the generated PTX using driver API, it errors out giving the reason as PTX alignment mismatch.

Upon investigating, I found that the bitcode generated by clang sees the CUDA vector types as structures instead of vectors which I believe is the main reason for this alignment mismatch. As you mentioned the alignment helps in selecting vectorized load/store instructions which has reasonable performance improvements. I agree this may not be a widely used use case as a user. We tried changing our representation from LLVM vector to structure type, but it has degraded performance severely for functions which are computationally not expensive. And using LLVM structure type has the drawback of creating an alloca to pass the value as argument to the function compiled by clang and results in more performance loss as it is being stored in the stack of the kernel instead of registers.

I hope this brief explanation along with previous sample code provide a good picture of our use case.

**Comment by rashedmyt** [comment]
I discovered #51566 when compiling the external CUDA device code we have with clang.

**Comment by rashedmyt** [comment]
Modifying your example to use CUDA vector types.
https://cuda.godbolt.org/z/eM35Y1nxc
Notice the load before call

**Comment by Artem-B** [comment]
I'm not sure this example is due to the alignment of return calls. AFACIT we've failed to vectorize the load of the kernel's `*in`. That's a problem, but it's likely caused by something else.

Can you by any chance post an example of the IR that results in the ptxas error you've mentioned? Or an excerpt of the generated PTX? I think I can see how ptxas may complain if it sees a vectorized load/store to/from something with an explicitly declared alignment that's too small. 

I guess for the result type we can lie and claim that it's always aligned by the size of the record/array element.

**Comment by rashedmyt** [comment]
The example in the previous comment was just to highlight the fact that using CUDA vector types generates different load/store instructions when compiled with NVCC and Clang and hence can have reasonable performance impact. It has nothing to do with the alignment. Apologies for the confusion.

I am attaching the global kernel IR, cuda file containing implementation of uplus, README file containing instructions and relevant intermediate LLVM IR files and the final PTX generated while following the instructions.
[final.zip](https://github.com/llvm/llvm-project/files/7830451/final.zip)

**Comment by Artem-B** [comment]
In your example the real issue is that uplus.cu produces 
```
%struct.double2.0 = type { double, double }
define dso_local %struct.double2.0 @uplus(%struct.double2.0* nocapture readonly byval(%struct.double2.0) align 16 %0) 
```
kernel.ll declares it  as `<2 x double> @uplus(<2 x double>)`, which is not quite the same.
The natural alignment of the vector is different of that of a struct and that's what leads to invalid PTX.

**Comment by rashedmyt** [comment]
Agreed, that is what I have mentioned in the comment containing brief overview of the use case. In CUDA, usage of double2 and other similar data structures are treated as vectors. But clang is not respecting that and treating them as structures.

**Comment by Artem-B** [comment]
Changing the alignment of the return value in PTX would be papering over the problem in this case.

Ideally, it would be great if CUDA headers would use something like `typedef double vd2 __attribute__((ext_vector_type(2)))` instead of a struct for its vector types, but that has its own issues -- e.g. the whole vector will be loaded/stored even when we access only one element, which is not the case for struct.

Alternatively, you could call `uplus` with correct signature and could use a `double2`->`<2xdouble>` conversion function. It would make sure that LLVM has correct information about the types used. E.g. something like this: https://godbolt.org/z/cr8hGffza

Regardless of that, considering that `.param` return values are special and are completely under NVPTX control, we should be free to align them as we want. It will not make much of a difference in practice, but it would bring us closer to what NVCC generates. I would still advise against intermixing struct and vector LLVM types, even if it happens to work sometimes.

**Comment by rashedmyt** [comment]
We are currently using the workaround you suggested i.e., we have conversion code in the JIT compiler to convert b/w vector and structure types. There was severe performance drop when we tried using structure types instead of vector types in the JIT compiler for computationally not expensive functions.. However, we were using LLVM 10 when the performance drop was observed. I have not checked since then. So things may have changed now.

**Comment by Artem-B** [comment]
If performance drop was attributed to local copies of byval arguments, you may want to try with a recent LLVM. LLVM at HEAD  should be able to avoid them in case it can prove that the argument is never modified. It's not perfect, but it's certainly much better than it used to be.

**Comment by Artem-B** [comment]
Upon closer look at the original issue (passing double2 args using f64 loads instead of v2.f64 loads), we should be able to deal with it. We are already vectorizing passing double2 to `add` and storing result to `output`. We do appear to have sufficient alignment info at least in some cases when we fail to vectorize loads. That's something we should fix.

**Comment by rashedmyt** [comment]
Part of the issue has been fixed with https://github.com/llvm/llvm-project/commit/66e8163f53cacc704aab9d4c81f208727e37d3d0.

The remaining issue is attributed to clang not generating alignment information for struct return types due to which `ld` from the `retval0` cannot be vectorized. Here is the reply from NVIDIA verbatim
> Unfortunately, the information required to make this vectorization possible is simply not in the upstream LLVM IR when it reaches the NVPTX backend. When compiling with nvcc, the frontend will attach an !"align" "nvvm.annotation" to such an over-aligned return type to communicate to the NVPTX backend what the alignment on the return type must be, allowing for vectorization. On the other hand, cuda clang doesn't attach any such information to the IR so we have to be conservative.
>
> Building the reproducer with llvm-project/main ToT clang shows the following PTX:
	{ // callseq 0, 0
	.param .align 16 .b8 	param0[16];
	.param .align 16 .b8 	param1[16];
	.param .align 8 .b8 	retval0[16];
	ld.param.b64 	%rd4, [_Z6kernelP7double2S0_S0__param_2];
	ld.v2.b64 	{%rd5, %rd6}, [%rd4];
	st.param.v2.b64 	[param1], {%rd5, %rd6};
	ld.v2.b64 	{%rd7, %rd8}, [%rd3];
	st.param.v2.b64 	[param0], {%rd7, %rd8};
	call.uni (retval0), add, (param0, param1);
	ld.param.b64 	%rd9, [retval0];
	ld.param.b64 	%rd10, [retval0+8];
	} // callseq 0
	.loc	1 6 10                          // ./test.cu:6:10
	st.global.v2.b64 	[%rd2], {%rd9, %rd10};
	.loc	1 7 1                           // ./test.cu:7:1
	ret;

**Comment by Artem-B** [comment]
@AlexMaclean has been working on function arg alignment. Perhaps he may have some ideas on how to bridge the gap.
